rule RNAseQC:
    input :
        __RNAseQC__input_bam
    log :
        __RNAseQC__logs
    output:
        dir = __RNAseQC__output,
        png = __RNAseQC__figure
    params:
        genome = config['RNAseQC']['ref'],
        gtf = config['RNAseQC']['gtf'],
        rRNA = config['RNAseQC']['BWArRNA']

    run:
        import os
        import pandas as pd
        import matplotlib.pyplot as pl
        pl.style.use('ggplot')

        #create the sample index
        for file in {input}

        #run RNAseQC
        shell("RNA-SeQC -o {output.dir}  -r {params.genome}  -s {input} -t {params.gtf}  -BWArRNA {params.rRNA}  2> {log}")

        #plot results
            if sum(df['rRNA rate']) == 0 :
                frames = [df['rRNA rate'] ,df['Exonic Rate'], df['Intronic Rate'], df['Intergenic Rate']]
            else :
                frames = [df['rRNA rate'], df['Exonic Rate'] - df['rRNA rate'], df['Intronic Rate'],
                          df['Intergenic Rate']]
            res = pd.concat(frames, axis=1)
            ##l√©gender correctement
            res.plot(kind='bar', stacked=True)
            pl.ylim([0,1])
            pl.legend(('rRNA rate', 'Exonic Rate', 'Intronic Rate', 'Intergenic Rate'), loc='lower center',
                        bbox_to_anchor=(0.5, +1.00), fancybox=True, shadow=True, ncol=4)
            pl.tight_layout()
            pl.savefig("Transcript-associated_Reads_metrics.svg")





