{% macro text(cutoff) %}
<h4>{{ _('General') }}</h4>
<p>
	<a href="http://lomereiter.github.io/sambamba/" title="Sambamba">Sambamba</a>
	{{ _('is used for coverage analysis based on read alignment.') }}
	{{ _('Duplicates, reads with mapping quality less than 10 as well as bases with a base quality less than 10 are not included in the analysis.') }}
	{{ _('Bases that overlap between read pairs are only counted once.') }}

	{{ _('The parts of the genome that has been analyzed') }}
	{{ _('is composed of all protein coding, exonic intervals') }}
	{{ _('referenced in the') }}
	<a href="http://www.ncbi.nlm.nih.gov/CCDS/CcdsBrowse.cgi">{{ _('CCDS database') }}</a>.

	{% if includes_splice_sites %}
		{{ _('Additionally, the data also includes spicing sites') }}
		({{ _('donator and acceptor positions') }},
		{{ splice_site_interval }}
		{{ _('bases on either of each exon') }}).
	{% endif %}
</p>

{% if config.CHANJO_PANEL %}
	<p>
		{{ _('Final coverage metrics were calculated for genes from the clinical panels') }}:
		<code>{{ config.CHANJO_PANEL_NAME }}</code>.
		{{ _('Note that estimations were made regarding coverage and completeness') }}
		{{ _('on the level of transcripts') }}.
	</p>
{% endif %}

<p>
	<strong>{{ _('Completeness') }}</strong>:
	{{ _('defined as the ratio of bases') }}
	{{ _('sequenced deeper than a specified cutoff') }},
	{{ _('e.g. 10x') }}.
</p>

<p>
	<strong>{{ _('Transcript coverage') }}</strong>:
	{{ _('defined as the ratio of transcripts that are fully covered') }},
	{{ _('i.e. with a completeness of') }} 100% {{ _('at') }} {{ cutoff }}x.
</p>

<p>
	<strong>{{ _('Gender prediction from data') }}</strong>:
	{{ _('calculated by comparing relative chromosome coverage') }}
	(X/Y).
	{{ _('Some reads randomly map to the Y chromosome') }}.
	{{ _('Therefore, a less than 10 fold difference is used to infer a "male" sample') }}.
</p>
{% endmacro %}
